typing/typedecl.ml

```diff
@@ -389,14 +490,38 @@ let transl_declaration env sdecl (id, uid) =
             (fun () -> make_cstr scstr)
         in
         let tcstrs, cstrs = List.split (List.map make_cstr scstrs) in
-          Ttype_variant tcstrs, Type_variant cstrs
+        let rep =
+          if unbox then
+            let layout =
+              Type_layout.of_layout_annotation ~default:Type_layout.any
+                layout_annotation
+            in
+            Variant_unboxed layout
+          else
+            (* We mark all arg layouts "any" here.  They are updated later,
+               after the circular type checks make it safe to check layouts. *)
+            Variant_boxed (
+              Array.of_list (List.map (fun cstr ->
+                match Types.(cstr.cd_args) with
+                | Cstr_tuple args ->
+                  Array.make (List.length args) Type_layout.any
+                | Cstr_record _ -> [| Type_layout.any |])
```

Comment:
We've gone back and forth on this, but the current story is that for (non `[@@unboxed]`) constructors with record arguments, we treat them here as having one argument (the record) with layout `value` (which is assigned in `update_decl_layout` where we update all the layouts in the kind).  Note that the argument to `Cstr_record` is a `label_declaration list`, and we do store the layout for each field there in the `ld_layout` field.

I think this could go either way.  My memory is that you Leo and I looked at this together last year and decided this was the best compromise based on where the information is used, but I don't remember the specific considerations that led us to that conclusion.  At some point soon we should do a complete re-assessment of what layout information is stored where in kinds - the current choices are just what was most expedient in the course of trying to get the prototype working, and should be rationalized.